### ISSUE TITLE
`clip_targets_to_zero` ignored when saving model results

### ISSUE THREAD

**Issue by vlahm** [neuralhydrology/neuralhydrology#49]
Hello @kratzert. In `BaseTester.evaluate`, the clipping of negative predictions to zero, i.e.

```
# clip negative predictions to zero, if variable is listed in config 'clip_target_to_zero'
if target_variable in self.cfg.clip_targets_to_zero:
    sim = xarray.where(sim < 0, 0, sim)
```
is performed on a subset of `results`, and is therefore not reflected in the dictionary (e.g. `test_results.p`) saved by `self._save_results(results, epoch)`. If this is intentional, can you help me understand?

Thank you,
Mike

**Comment by kratzert** [comment]
Hi @vlahm,

thanks for your question. You are indeed correct, that this is intentional (but that doesn't it is the best thing to do :grimacing:).

We did this, because we wanted to store the raw model outputs, so that you can do whatever different post-processing you want, without re-running the evaluation. Note: The argument `clip_targets_to_zero` is only used when computing metrics and not e.g. during the training. So the model gets a gradient signal to _learn_ that it should not go below zero, rather than killing all gradients that come from "wrong" negative predictions. 
The main purpose and reason why we included this option is to be able to compute metrics on the validation set during the training (i.e. in-between training epochs) that reflect the values that we would compute after training with the most common post-processing that we apply (which is clip negative predictions to zero). However, in other studies you might be interested in doing a different post-processing (e.g. using the model outputs in a separate model for data assimilation). Maybe there it makes sense to use the raw outputs?! In this case, you would need to change the config after training, then re-run the evaluation. Possible, not sure if better.

To be honest, I don't have a strong opinion on that and could also see reasons why it would be better to store the results with the clipped values. @gauchm and @danklotz what do you think?

**Comment by gauchm** [comment]
I can see that this behavior might be surprising, but I would keep the logic as it is, storing the raw results rather than clipped values. My main reason is that it's easy to clip the values again if needed, but impossible to retrieve the raw predictions once they are clipped.

Like @kratzert says, it may also be a use-case to calculate clipped metrics during validation but then use non-clipped raw predictions for some downstream task, so our current solution seems the most flexible to me.

Still, I think we should at least make the behavior more explicit in the [docs entry](https://neuralhydrology.readthedocs.io/en/latest/usage/config.html#data-settings) to make people more aware of what the config option actually does.

**Comment by vlahm** [comment]
I agree that the flexibility of this logic is worth keeping, and that making it more explicit in the docs would clear up any confusion. Thanks for your answers!

**Comment by kratzert** [comment]
> I agree that the flexibility of this logic is worth keeping, and that making it more explicit in the docs would clear up any confusion. Thanks for your answers!

Okay, happy we could clear the confusion. I will update the docs and we will push it online with the next update we release here in the public repository.
